### ISSUE TITLE
Redundant functions between `Tensorflow Features Columns` API and `TF Keras Preprocessing Layers`

### ISSUE THREAD

**Issue by 00krishna** [tensorflow/tensorflow#35680]
In working with TF 2.0, I noticed that the TF Feature Columns API seems to overlap with the Keras Preprocessing layers (or Keras Utils). For example, you can create an `tf.feature_column.indicator_column()` which creates a bunch of dummy variables or a one-hot encoded matrix based upon a categorical variable. With Keras preprocessing or Utils, you can use the `tf.keras.backend.one_hot()` function to perform the same operation. I think there are similar overlaps between TF Feature Columns like the embedding columns and similar Keras functions for embedding columns.

I was just wondering if the steering committees for Tensorflow have any direction on whether they plan to promote one set of functions versus the other? Are there any plans to deprecate one set versus the other. For me, it is just a question of where to invest time and planning for code that might have to change in the future. Seems like maintaining the redundancy in the package will potentially lead to performance differences between similar functions, or confusion in setting up the code, etc. 

**NB** 
Oh yes, I actually asked this question in the Tensorflow Discussion forum, but no one answered it. @dynamicwebpaige even forwarded the message to @karmel and Mark Omernick, but no one responded. Hence, I posted here.

**Comment by 00krishna** [comment]
Hey Folks, any additional thoughts on this questions?

**Comment by 00krishna** [comment]
@fchollet I was wondering if you had any additional thoughts on this question? Seems like there should be a direction on what TF will go in terms of Features API versus the Keras preprocessing functions. Or does Google and the Community want to keep both sets of functions, even though they overlap.

**Comment by 00krishna** [comment]
@fchollet I thought I would just follow up on this question. Is there any direction on whether users should use the Feature Columns API, or the TF Keras Preprocessing features? Both seem to do very similar things, so I image that Google would not want to maintain redundant code. 
Thanks.

**Comment by ilango100** [comment]
According to this [Accepted RFC](https://github.com/tensorflow/community/blob/0650b34a755e51050ae4f0e91a4e773b2515b46f/rfcs/20191212-keras-categorical-inputs.md), the Keras Preprocessing Layers are introduced to replace feature columns and DenseFeatures. So, I guess once those layers are ready and exported, the plan is to deprecate feature columns. As mentioned in RFC, it is difficult to use feature columns in Keras functional API.

Preprocessing layers are better than feature columns in lot ways. Using layers like Normalization, TextVectorization avoids building preprocessing pipeline during prediction time, since they are built right into the model.

But I'm concerned that preprocessing layers like [RandomCrop](https://www.tensorflow.org/versions/r2.2/api_docs/python/tf/keras/layers/experimental/preprocessing/RandomCrop), [RandomRotation](https://www.tensorflow.org/versions/r2.2/api_docs/python/tf/keras/layers/experimental/preprocessing/RandomRotation), which are required only at training time are redundant. Since they are required only at training time and not at prediction time, it makes more sense to apply those preprocessing steps in tf.data.Dataset instead of in the model.

```
dataset = ... # Load images
dataset.map(lambda x: tf.image.random_crop(x, (size))
dataset.map(tf.image.random_flip_up_down)

model = ... #Build model

model.fit(dataset)

# Prediction
model.prediction(..)
```
So in my view, the criteria to define Preprocessing layers vs Dataset operations can be:
- If the step is required at both training and prediction time -> Preprocessing layer
- If the step is required only during training -> No Preprocessing layer, use Dataset operation

This can speed up the inference by eliminating unnecessary layer.

**Comment by 00krishna** [comment]
@ilango100 Thanks for the info. Yeah, I saw some mentions of the Keras Preprocessing layers but I think only a few of them are available under experimental right now. I agree that the feature layers are a bit confusing, but I have finally become used to them. That is good though, now I know what to keep my eye on for the future.

**Comment by goldiegadde** [comment]
@00krishna  Keras Preprocesing layers are released in TF 2.3.0 release. Check the [release notes](https://github.com/tensorflow/tensorflow/releases/tag/v2.3.0) for details. 
In addition you can also checkout the FC to Preprocessing Layers [migration guide](https://github.com/tanzhenyu/community/blob/5507ba06c8f43830f4a77ea85bc8d345f5b60765/rfcs/20191212-keras-categorical-inputs.md)

"Feature columns will continue to work for canned Estimators, but we believe Keras Preprocessing Layers better reflect the TensorFlow v2 programming models, and they integrate better with Keras, which is the preferred high level API in TF2. Thus, we encourage you to use Keras Preprocessing Layers for your preprocessing needs, and feature columns will be deprecated at some point in the future when the use cases can be fully replaced by Preprocessing Layers + Keras models."

**Comment by goldiegadde** [comment]
@00krishna am closing this based on my previous comment, if you have any usecase that is not satisfied with the Keras Preprocessing Layers, please feel free to open a new issue.
